File: ieee2022/java/src/Graphs/NormalGraphs/Graph.java
package Graphs.NormalGraphs;

import java.util.*;
import java.util.AbstractMap.SimpleImmutableEntry;

public class Graph<T, D> {

    public HashMap<T, Node<T, D>> nodes;

    public Graph() {
        this.nodes = new HashMap<>();
    }

    public void addNode(T key, D data) {
        this.nodes.put(key, new Node<>(key, data));
    }

    public void addNode(T key) {
        this.nodes.put(key, new Node<>(key, null));
    }

    public void addConnection(T source, T target, int distance) {
        this.nodes.get(source).addNeighbour(this.nodes.get(target), distance);
    }

    public void addConnection(T source, T target) {
        this.nodes.get(source).addNeighbour(this.nodes.get(target), 1);
    }

    public void addBidirectionalConnection(T a, T b, int distance) {
        this.nodes.get(a).addNeighbour(this.nodes.get(b), distance);
        this.nodes.get(b).addNeighbour(this.nodes.get(a), distance);
    }

    public void addBidirectionalConnection(T a, T b) {
        this.nodes.get(a).addNeighbour(this.nodes.get(b), 1);
        this.nodes.get(b).addNeighbour(this.nodes.get(a), 1);
    }

    public void addConnectionWithObject(Node<T, D> source, Node<T, D> target, int distance) {
        source.addNeighbour(target, distance);
    }

    public void addConnectionWithObject(Node<T, D> source, Node<T, D> target) {
        source.addNeighbour(target, 1);
    }

    public void addBidirectionalConnectionWithObject(Node<T, D> a, Node<T, D> b, int distance) {
        a.addNeighbour(b, distance);
        b.addNeighbour(a, distance);
    }

    public void addBidirectionalConnectionWithObject(Node<T, D> a, Node<T, D> b) {
        a.addNeighbour(b, 1);
        b.addNeighbour(a, 1);
    }

    public SimpleImmutableEntry<HashMap<T, Integer>, HashMap<T, T>> dijkstra(T start) {
        Node<T, D> startNode = this.nodes.get(start);
        HashMap<T, Integer> distances = new HashMap<>();
        HashMap<T, T> previous = new HashMap<>();
        PriorityQueue<SimpleImmutableEntry<T, Integer>> queue = new PriorityQueue<>(11, new Comparator<SimpleImmutableEntry<T, Integer>>() {
            @Override
            public int compare(SimpleImmutableEntry<T, Integer> o1, SimpleImmutableEntry<T, Integer> o2) {
                return o1.getValue().compareTo(o2.getValue());
            }
        });
        HashSet<T> explored = new HashSet<>();
        for (T key: this.nodes.keySet()) {
            distances.put(key, Integer.MAX_VALUE);
            previous.put(key, null);
        }
        distances.put(start, 0);
        queue.add(new SimpleImmutableEntry<>(start, 0));
        while (!queue.isEmpty()) {
            SimpleImmutableEntry<T, Integer> entry = queue.poll();
            if (explored.contains(entry.getKey())) {
                continue;
            }
            explored.add(entry.getKey());
            for (SimpleImmutableEntry<Integer, Node<T, D>> neighbour: this.nodes.get(entry.getKey()).neighbours) {
                int altDistance = entry.getValue() + neighbour.getKey();
                if (altDistance < distances.get(neighbour.getValue().key)) {
                    distances.put(neighbour.getValue().key, altDistance);
                    previous.put(neighbour.getValue().key, entry.getKey());
                    queue.add(new SimpleImmutableEntry<>(neighbour.getValue().key, altDistance));
                }
            }
        }
        return new SimpleImmutableEntry<>(distances, previous);
    }

    public static void main(String[] args) {
        long timeTaken = 0;
        int nodeAmount = 1000;
        int connectionAmount = 3000;
        Random random = new Random();
        for (int i = 0; i < 10000; i++) {
            Graph<String, Integer> graph = new Graph<>();
            for (int j = 0; j < nodeAmount; j++) {
                graph.addNode("A" + j, j);
            }
            ArrayList<Node<String, Integer>> nodes = new ArrayList<>(graph.nodes.values());
            for (int j = 0; j < connectionAmount; j++) {
                Node<String, Integer> a = nodes.get(random.nextInt(nodes.size() - 1));
            }
            timeTaken -= System.currentTimeMillis();


            timeTaken += System.currentTimeMillis();
        }
        System.out.println(timeTaken);
    }
}

File: ieee2022/python/graphs/normal_graphs/graph.py
import heapq
from collections import deque

from node import Node


class Graph:
    def __init__(self):
        self.nodes = {}

    def add_node(self, key, data=None):
        self.nodes[key] = Node(key, data)

    def add_connection(self, source, target, distance=1):
        self.nodes[source].add_neighbour(self.nodes[target], distance)

    def add_bidirectional_connection(self, a, b, distance=1):
        self.nodes[a].add_neighbour(self.nodes[b], distance)
        self.nodes[b].add_neighbour(self.nodes[a], distance)

    def add_connection_with_object(self, source, target, distance=1):
        source.add_neighbour(target, distance)

    def add_bidirectional_connection_with_object(self, a, b, distance=1):
        a.add_neighbour(b, distance)
        b.add_neighbour(a, distance)

    def dijkstra(self, start):
        start_node = self.nodes[start]
        dists, prev, queue, explored = {}, {}, [], set()
        for key, node in self.nodes.items():
            dists[key] = float("inf")
            prev[key] = None
        dists[start] = 0
        heapq.heappush(queue, (0, start))
        while queue:
            dist, key = heapq.heappop(queue)
            if key in explored:
                continue
            explored.add(key)
            for dist2, neighbour in self.nodes[key].neighbours:
                alt_dist = dist + dist2
                if alt_dist < dists[neighbour.key]:
                    dists[neighbour.key] = alt_dist
                    prev[neighbour.key] = key
                    heapq.heappush(queue, (alt_dist, neighbour.key))
        return dists, prev

    def dijkstra_with_target(self, start, target):
        start_node = self.nodes[start]
        dists, prev, queue, explored = {}, {}, [], set()
        for key, node in self.nodes.items():
            dists[key] = float("inf")
            prev[key] = None
        dists[start] = 0
        heapq.heappush(queue, (0, start))
        while queue:
            dist, key = heapq.heappop(queue)
            if key in explored:
                continue
            if key == target:
                return dists, prev, True
            explored.add(key)
            for dist2, neighbour in self.nodes[key].neighbours:
                alt_dist = dist + dist2
                if alt_dist < dists[neighbour.key]:
                    dists[neighbour.key] = alt_dist
                    prev[neighbour.key] = key
                    heapq.heappush(queue, (alt_dist, neighbour.key))
        return dists, prev, False

    @staticmethod
    def get_path(source, target, dists, prev):
        path = []
        current = target
        while current != source:
            if prev[current] is None:
                return None
            path.append(current)
            current = prev[current]
        path.append(source)
        return path[::-1], dists[target]

    def dfs(self, start):
        visited = set()
        stack = deque()
        stack.append(start)
        while stack:
            next_key = stack.pop()
            if next_key in visited:
                continue
            print(f"Visiting {next_key}")
            visited.add(next_key)
            for _, neighbour in self.nodes[next_key].neighbours:
                stack.append(neighbour.key)


if __name__ == '__main__':
    graph = Graph()
    graph.add_node("A")
    graph.add_node("B")
    graph.add_node("C")
    graph.add_node("D")
    graph.add_node("E")
    graph.add_bidirectional_connection("A", "B", 1)
    graph.add_bidirectional_connection("A", "C", 2)
    graph.add_bidirectional_connection("B", "D", 10)
    graph.add_bidirectional_connection("C", "D", 4)
    graph.add_bidirectional_connection("B", "C", 2)
    dists, prev, success = graph.dijkstra_with_target("A", "D")
    print(f"Path from 'A' to 'D': {Graph.get_path('A', 'D', dists, prev)}")

    print("Doing DFS")
    graph.dfs("B")
